lib/paratii.eth.vouchers.js

```diff
@@ -0,0 +1,63 @@
+import { getInfoFromLogs } from './utils.js'
+let dopts = require('default-options')
+
+export class ParatiiEthVouchers {
+  constructor (context) {
+    // context is a ParatiiEth instance
+    this.eth = context
+  }
+
+  async getVouchersContract () {
+    let contract = await this.eth.getContract('Vouchers')
+    if (contract.options.address === '0x0') {
+      throw Error('There is not Vouchers contract known in the registry')
+    }
+    return contract
+  }
+
+  async add (options) {
+    let defaults = {
+      voucherCode: null,
+      amount: undefined
+    }
+    options = dopts(options, defaults)
+
+    if (options.voucherCode === null) {
+      let msg = `Voucher Code argument must not be null`
+      throw Error(msg)
+    }
+    if (typeof options.amount !== 'number') {
+      let msg = `Amount argument needs to be a number`
+      throw Error(msg)
+    }
+    if (typeof options.amount !== 'number') {
+      let msg = `Amount needs to be greater than zero`
+      throw Error(msg)
+    }
+
+    let contract = await this.getVouchersContract()
+    let hashedVoucher = await contract.methods.hashVoucher(options.voucherCode).call()
+    let tx = await contract.methods.create(
+      hashedVoucher,
+      options.amount
+    ).send()
+    let voucherId = getInfoFromLogs(tx, 'LogCreateVoucher', '_hashedVoucher')
+
+    return voucherId
+  }
+
+  async redeem (voucherCode) {
+    let contract = await this.getVouchersContract()
+    let tx = await contract.methods.redeem(voucherCode).send()
+    try {
+      let claimant = getInfoFromLogs(tx, 'LogRedeemVoucher', '_claimant')
+      if (claimant === this.eth.config.accounts.address) {
+        return true
+      } else {
+        return false
+      }
+    } catch (e) {

```

Comment:
This error will be pretty much useless, because the EVM (ethereum virutal machine) returns very non-specific error messages. What may be a good idea is that, on error, the code looks for some probable causes. Here are some:

* the contract does not have enough PTI to pay the claimant
* the voucher exists but was already used
* the user is trying to claim an non-existing voucher (i.e. hash(voucherCode) is not a voucher in contract.vouchers()

Reply:
I'm working on that, I'm not quite clear on how to implement the checks...

Reply:
- for the "contract has enough funds" you can check like this: https://github.com/Paratii-Video/paratii-contracts/blob/dev/test/voucher.js#L30
- for the "exists but was already used", you could check `(await vouchers.vouchers(hashedVoucher))[0]` is an ethereum address (this means that the voucher was paid)
- Non-existing vouchers: `(await vouchers.vouchers(hashedVoucher))[1]` (which is the amount of the vocher) is `0`. 

(The Solidity/EVM mapping types are a bit weird: basically, the values are always initialize dwith the default values of their types. So, if you have a mapping from, say, integers to integers called `foo`, then `foo[x]` will return `0` instead of a more useful `null` or `undefined`, even when it was never initialized explicitly.. )